### PR TITLE
test(contract): integration test to ensure contract rejects invalid remote attestations

### LINKED ISSUE #727: The contract should reject invalid remote attestations.
### Background

Check each one of the verification methods in the contract work as expected:
https://github.com/near/mpc/blob/7e129e8d6e56402aa7ccf52087fd9b9de5e093d1/libs/chain-signatures/contract/src/tee/tee_participant.rs#L122-L149

Some test cases:
- Random bytes.
- Change some bytes of a valid quote.
- MPC node is running a different version and generates a signature. It should be rejected as the MPC version is not allowed.

### User Story

_No response_

### Acceptance Criteria

We can Create a malicious version of the MPC code that tampers with the quotes to check that the contract rejects these.

or/and

Create unit/integration tests for the contract that covers these verification methods and edge cases.

### Resources & Additional Notes

_No response_

### PR DESCRIPTION
**Pull request by pbeza**
Fixes #727

### CODE REVIEW

File: libs/chain-signatures/contract/src/tee/proposal.rs [hunk 1]
```diff
@@ -133,25 +133,9 @@ impl AllowedDockerImageHashes {
...
-environment:
-  - DOCKER_CONTENT_TRUST=1
-  - DEFAULT_IMAGE_DIGEST=sha256:{}
-volumes:
-  - /var/run/docker.sock:/var/run/docker.sock
-  - /var/run/dstack.sock:/var/run/dstack.sock
-  - /tapp:/tapp:ro
-  - /var/lib/docker/volumes/shared-volume/_data:/mnt/shared:ro
-"#,
+        let filled_yaml = format!("version: '3.8'\n\nservices:\n  launcher:\n    image: barakeinavnear/launcher@sha256:1ea7571baf18bd052359abd2a1f269e7836f9bad2270eb55fc9475aa327f8d96\n\n # isuse #531: TODO (security): Replace with a specific image digest\n    container_name: launcher\n\n    environment:\n      - DOCKER_CONTENT_TRUST=1\n      - DEFAULT_IMAGE_DIGEST=sha256:{}\n\n    volumes:\n      - /var/run/docker.sock:/var/run/docker.sock\n      - /var/run/dstack.sock:/var/run/dstack.sock\n      - /tapp:/tapp:ro\n      - shared-volume:/mnt/shared:ro\n\n    security_opt:\n      - no-new-privileges:true\n\n    read_only: true\n\n    tmpfs:\n      - /tmp\n\nvolumes:\n  shared-volume:\n    name: shared-volume",
```
[the remark this answers is not in the document]
Comment (pbeza) [reply libs/chain-signatures/contract/src/tee/proposal.rs:136]: 🤷🏼

File: libs/chain-signatures/contract/src/tee/tee_state.rs [hunk 2]
```diff
@@ -45,7 +45,7 @@ impl TeeState {
         current_time_milliseconds / 1_000
     }
 
-    /// May return an error
+    /// May return an error // TODO not true
```
[the remark this answers is not in the document]
Comment (pbeza) [reply libs/chain-signatures/contract/src/tee/tee_state.rs]: Fixed: https://github.com/near/mpc/pull/828/commits/3681e1da9950360e73716822935e136a4f02f203.

File: attestation/src/attestation.rs [hunk 3]
```diff
@@ -50,6 +52,35 @@ pub struct DstackAttestation {
...
+            .field("collateral", &truncate_debug(&self.collateral, MAX_BYTES))
+            .field("tcb_info", &truncate_debug(&self.tcb_info, MAX_BYTES))
+            .field(
+                "expected_measurements",
+                &truncate_debug(&self.expected_measurements, MAX_BYTES),
+            )
+            .finish()
+    }
+}
+
```
Comment (pbeza) [inline attestation/src/attestation.rs:83]: FWIW, if you comment this out (falling back to the default Debug trait implementation) and then run:
```
cargo test test_submit_participant_info_succeeds_with_valid_tee_attestation
```
you’ll get:
```
ExecutionFinalResult { total_gas_burnt: NearGas { inner: 8453505019422 }, transaction: ExecutionOutcome { transaction_hash: B595ocNNrTXSru2HA4aTPkZsc8S1mnBQCyTzJsxDHScF, block_hash: 5AT1ibp3jUHDfsxhCw97CYMgrp29YRqyZ8RgU83ceWgY, logs: [], receipt_ids: [5UZUnsWmudMFubH6WhAq2G7RX3KnjWpxxf966XryhiV], gas_burnt: NearGas { inner: 1473830964320 }, tokens_burnt: NearToken { inner: 147383096432000000000 }, executor_id: AccountId("dev-20250822120347-56275495132198.test.near"), status: SuccessReceiptId(5UZUnsWmudMFubH6WhAq2G7RX3KnjWpxxf966XryhiV) }, receipts: [ExecutionOutcome { transaction_hash: 5UZUnsWmudMFubH6WhAq2G7RX3KnjWpxxf966XryhiV, block_hash: FDZehqCuAjFPqjq943DJS6mMegJpwep21qE66hGjvWCE, logs: [], receipt_ids: [2Jzj4hXmaGRdQqXuXHoyX1CPWZZLvYN1zoow6DgwDPTT], gas_burnt: NearGas { inner: 6979674055102 }, tokens_burnt: NearToken { inner: 697967405510200000000 }, executor_id: AccountId("dev-20250822120345-17791131168986.test.near"), status: Failure(ActionError(ActionError { index: Some(0), kind: FunctionCallError(ExecutionError("The length of a log message 53888 exceeds the limit 16384")) })) }], status: Failure(ActionError(ActionError { index: Some(0), kind: FunctionCallError(ExecutionError("The length of a log message 53888 exceeds the limit 16384")) })) }
```

Comment (DSharifi) [reply attestation/src/attestation.rs:83]: Nice to catch these things with the test.

Comment (DSharifi) [reply attestation/src/attestation.rs:83]: Do we know which field(s) that are the problematic ones and need truncation? Instead of truncating, we could try to create a compact summary.

If it's the `QuoteBytes` I wonder if we can display a hash instead of all the bytes (it's order of thousands bytes IIRC). Also the `TcbInfo` struct is probably gigantic with all it's fields. Maybe we don't need to display everything there as well.

File: attestation/src/attestation.rs [hunk 4]
```diff
@@ -129,8 +160,8 @@ impl Attestation {
             )
     }
 
-    /// Replays RTMR3 from the event log by hashing all relevant events together and verifies all digests
-    /// are correct
+    /// Replays RTMR3 from the event log by hashing all relevant events together and verifies all
+    /// digests are correct
```
Comment (DSharifi) [inline attestation/src/attestation.rs:164]: It's fine for this and the previous PR, but I'd prefer that you turn off your config for comment wrapping to avoid these diffs OR we make it enforced on CI so it's standardized.

The latter might be tricky as @gilcu3 discovered the `comment_width` attribute for `rustfmt` is not stabilisied and it was removed in https://github.com/near/mpc/pull/790

File: attestation/Cargo.toml [hunk 5]
```diff
@@ -24,6 +24,8 @@ tracing = { workspace = true }
 rstest = { workspace = true }
 
 [features]
+default = ["test-utils"]
+test-utils = []
```
Comment (DSharifi) [inline attestation/Cargo.toml:28]: I'd like to not have `test-utils` part of the default feature set, so it can be specifically opted in for the dev dependencies. Now it's activated for production code.

The most ideal way would be to have a more general `test-utils` crate (not module) which need not only contains utils for `attestation` but also other crates in the future.

Then we'd only depend on `test-utils` crate in `[dev-dependencies]`, and flag it as a test target only.

File: libs/chain-signatures/contract/tests/tee.rs [hunk 6]
```diff
@@ -230,4 +232,62 @@ async fn get_participants(contract: &Contract) -> Result<usize> {
...
 }
 
-// todo [#514](https://github.com/near/mpc/issues/514)
+/// Tests TEE attestation validation including both positive and negative cases.
+/// Verifies that:
+/// 1. Attestation fails when no MPC hash is approved
+/// 2. Attestation succeeds with valid TLS key and approved MPC hash
+/// 3. Attestation fails when TLS key doesn't match the one in report data
+#[tokio::test]
+async fn test_tee_attestation_validation_with_valid_and_invalid_tls_keys() -> Result<()> {
```
Comment (DSharifi) [inline libs/chain-signatures/contract/tests/tee.rs:241]: Can you split these three test cases into three tests?

**Review by DSharifi — CHANGES_REQUESTED** [review]
Thanks for the good work, love that these tests are catching bugs such as the deserialization.

Can you break up the test cases, and address the `test-utils` feature flag comment?

File: libs/chain-signatures/contract/tests/tee.rs (hunk 6 above)
Comment (pbeza) [reply libs/chain-signatures/contract/tests/tee.rs]: Sure: https://github.com/near/mpc/pull/828/commits/f656a30d6dee6548b9dc8021825b90f24584336e. ✅

File: attestation/src/attestation.rs (hunk 3 above)
Comment (pbeza) [reply attestation/src/attestation.rs:83]: For the attestation mockup, we’re using:

quote 18759 bytes
collateral: 20501 bytes
tcb_info: 6879 bytes

```
{
    quote: "QuoteBytes([4, 0, 2, 0, 129, 0, 0, 0, 0, 0, 0, 0, 147, 154, 114, 51, 247, 156, 76, 169, 148, 10, 13, 179, 149, 127, 6, 7, 128, 52, 244, 195, 150, 62, 198, 18, 157, 233, 20, 30, 146, 118, 66, 119, 0, 0, 0, 0, 7, 1, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 73, 182, 111, 170, 69, 29, 25, 235, 189, 190, 137, 55, 27, 141, 175, 43, 101, 170, 57, 132, 236, 144, 17, 3, 67, 233, 226, 238, 193, 22, 175, 8, 133, 15, 162, 14, 59, 26, 169, 168, 116, 215, 122, 101, 56, 14, 231, 230, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 16, 0, 0, 0, 0, 231, 2, 6, 0, 0, 0, 0, 0, 198, 133, 24, 160, 235, 180, 33, 54, 193, 43, 34, 117, 22, 79, 140, 114, 242, 95, 169, 163, 67, 146, 34, 134, 135, 237, 110, 156, 174, 185, 192, 241, 219, 216, 149, 233, 207, 71, 81, 33, 192, 41, 220, 71, 231, 14, 145, 253, 1, 46, 100, 145, 24, 148, 136, 118, 206, 127, 35, 144, 233, 152, 5, 96, 50, 120, 63, 191, 8, 130, 114, 26, 70, 78, 212, 40, 137, 107, 61, 133, 148, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 55, 68, 177, 84, 6, 149, 0, 164, 102, 245, 20, 37, 59, 73, 133, 130, 153, 178, 225, 189, 196, 78, 61, 85, 115, 55, 216, 30, 130, 139, 237, 246, 160, 65, 15, 39, 211, 161, 140, 147, 46, 94, 73, 225, 196, 33, 87, 55, 75, 102, 232, 136, 200, 223, 167, 165, 4, 252, 124, 160, 96, 171, 158, 45, 5, 18, 51, 241, 21, 215, 19, 4, 8, 85, 112, 199, 172, 113, 245, 161, 144, 163, 226, 55, 209, 95, 9, 101, 150, 122, 120, 83, 155, 160, 215, 135, 90, 65, 201, 247, 28, 229, 101, 91, 107, 166, 5, 254, 13, 0, 160, 160, 90, 221, 116, 113, 172, 170, 166, 170, 21, 91, 206, 30, 4, 184, 32, 79, 15, 255, 174, 194, 230, 201, 95, 252, 20, 66, 179, 126, 20, 17, 39, 217,... (truncated 18759 bytes)",
    collateral: "Collateral(QuoteCollateralV3 { pck_crl_issuer_chain: \"-----BEGIN CERTIFICATE-----\\nMIICljCCAj2gAwIBAgIVAJVvXc29G+HpQEnJ1PQzzgFXC95UMAoGCCqGSM49BAMC\\nMGgxGjAYBgNVBAMMEUludGVsIFNHWCBSb290IENBMRowGAYDVQQKDBFJbnRlbCBD\\nb3Jwb3JhdGlvbjEUMBIGA1UEBwwLU2FudGEgQ2xhcmExCzAJBgNVBAgMAkNBMQsw\\nCQYDVQQGEwJVUzAeFw0xODA1MjExMDUwMTBaFw0zMzA1MjExMDUwMTBaMHAxIjAg\\nBgNVBAMMGUludGVsIFNHWCBQQ0sgUGxhdGZvcm0gQ0ExGjAYBgNVBAoMEUludGVs\\nIENvcnBvcmF0aW9uMRQwEgYDVQQHDAtTYW50YSBDbGFyYTELMAkGA1UECAwCQ0Ex\\nCzAJBgNVBAYTAlVTMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAENSB/7t21lXSO\\n2Cuzpxw74eJB72EyDGgW5rXCtx2tVTLq6hKk6z+UiRZCnqR7psOvgqFeSxlmTlJl\\neTmi2WYz3qOBuzCBuDAfBgNVHSMEGDAWgBQiZQzWWp00ifODtJVSv1AbOScGrDBS\\nBgNVHR8ESzBJMEegRaBDhkFodHRwczovL2NlcnRpZmljYXRlcy50cnVzdGVkc2Vy\\ndmljZXMuaW50ZWwuY29tL0ludGVsU0dYUm9vdENBLmRlcjAdBgNVHQ4EFgQUlW9d\\nzb0b4elAScnU9DPOAVcL3lQwDgYDVR0PAQH/BAQDAgEGMBIGA1UdEwEB/wQIMAYB\\nAf8CAQAwCgYIKoZIzj0EAwIDRwAwRAIgXsVki0w+i6VYGW3UF/22uaXe0YJDj1Ue\\nnA+TjD1ai5cCICYb1SAmD5xkfTVpvo4UoyiSYxrDWLmUR4CI9NKyfPN+\\n-----END CERTIFICATE-----\\n-----BEGIN CERTIFICATE-----\\nMIICjzCCAjSgAwIBAgIUImUM1lqdNInzg7SVUr9QGzknBqwwCgYIKoZIzj0EAwIw\\naDEaMBgGA1UEAwwRSW50ZWwgU0dYIFJvb3QgQ0ExGjAYBgNVBAoMEUludGVsIENv\\ncnBvcmF0aW9uMRQwEgYDVQQHDAtTYW50YSBDbGFyYTELMAkGA1UECAwCQ0ExCzAJ\\nBgNVBAYTAlVTMB4XDTE4MDUyMTEwNDUxMFoXDTQ5MTIzMTIzNTk1OVowaDEaMBgG\\nA1UEAwwRSW50ZWwgU0dYIFJvb3QgQ0ExGjAYBgNVBAoMEUludGVsIENvcnBvcmF0\\naW9uMRQwEgYDVQQHDAtTYW50YSBDbGFyYTELMAkGA1UECAwCQ0ExCzAJBgNVBAYT\\nAlVTMFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEC6nEwMDIYZOj/iPWsCzaEKi7\\n1OiOSLRFhWGjbnBVJfVnkY4u3IjkDYYL0MxO4mqsyYjlBalTVYxFP2sJBK5zlKOB\\nuzCBuDAfBgNVHSMEGDAWgBQiZQzWWp00ifODtJVSv1AbOScGrDBSBgNVHR8ESzBJ\\nMEegRaBDhkFodHRwczovL2NlcnRpZmljYXRlcy50cnVzdGVkc2VydmljZXMuaW50\\nZWwuY29tL0ludGVsU0dYUm9vdENBLmRlcjAdBgNVHQ4EFgQUImUM1lqdNInzg7SV\\nUr9QGzknBqwwDgYDVR0PAQH/BAQDAgEGMBIGA1UdEwEB/wQIMAYBAf8CAQEwCgYI\\nKoZIzj0EAwIDSQAwRgIhAOW/5QkR+S9CiSDcNoowLuPRLsWGf/Yi7GSX94BgwTwg\\nAiEA4J0lrHoMs+Xo5o/sX6O9QWxHRAvZUGOdRQ7cvqRXaqI=\\n-----END CERTIFICATE-----\\n\", root_ca_crl: [48, 130, 1, 32, 48, 129, 200, 2, 1, 1, 48... (truncated 20501 bytes)",
    tcb_info: "TcbInfo { mrtd: \"c68518a0ebb42136c12b2275164f8c72f25fa9a34392228687ed6e9caeb9c0f1dbd895e9cf475121c029dc47e70e91fd\", rtmr0: \"3744b154069500a466f514253b49858299b2e1bdc44e3d557337d81e828bedf6a0410f27d3a18c932e5e49e1c4215737\", rtmr1: \"4b66e888c8dfa7a504fc7ca060ab9e2d051233f115d71304085570c7ac71f5a190a3e237d15f0965967a78539ba0d787\", rtmr2: \"5a41c9f71ce5655b6ba605fe0d00a0a05add7471acaaa6aa155bce1e04b8204f0fffaec2e6c95ffc1442b37e141127d9\", rtmr3: \"74c3f1e616efcab376d473c9a1060ee40ecae125c17d468c01bbaf87f26e8c60208d335513738542c99841a1471aa5cb\", os_image_hash: \"\", compose_hash: \"2e649118948876ce7f2390e998056032783fbf0882721a464ed428896b3d8594\", device_id: \"f61bdbd417b4a1b6519e698e2ea90420c21140702b608c08c72fb51ec52a5fea\", app_compose: \"{\\n    \\\"manifest_version\\\": 2,\\n    \\\"name\\\": \\\"launcher_test_app\\\",\\n    \\\"runner\\\": \\\"docker-compose\\\",\\n    \\\"docker_compose_file\\\": \\\"version: '3.8'\\\\n\\\\nservices:\\\\n  launcher:\\\\n    image: barakeinavnear/launcher@sha256:1ea7571baf18bd052359abd2a1f269e7836f9bad2270eb55fc9475aa327f8d96\\\\n\\\\n # isuse #531: TODO (security): Replace with a specific image digest\\\\n    container_name: launcher\\\\n\\\\n    environment:\\\\n      - DOCKER_CONTENT_TRUST=1\\\\n      - DEFAULT_IMAGE_DIGEST=sha256:c22901e52cfa91b2e71eb8694ac9558065c6e3b13783d9e3d36b792d93ce153b\\\\n\\\\n    volumes:\\\\n      - /var/run/docker.sock:/var/run/docker.sock\\\\n      - /var/run/dstack.sock:/var/run/dstack.sock\\\\n      - /tapp:/tapp:ro\\\\n      - shared-volume:/mnt/shared:ro\\\\n\\\\n    security_opt:\\\\n      - no-new-privileges:true\\\\n\\\\n    read_only: true\\\\n\\\\n    tmpfs:\\\\n      - /tmp\\\\n\\\\nvolumes:\\\\n  shared-volume:\\\\n    name: shared-volume\\\",\\n    \\\"kms_enabled\\\": false,\\n    \\\"gateway_enabled\\\": false,\\n    \\\"local_key_provider_enabled\\\": true,\\n    \\\"key_provider_id\\\": \\\"\\\",\\n    \\\"public_logs\\\": true,\\n    \\\"public_sysinfo\\\": true,\\n    \\\"allowed_envs\\\": [],\\n    \\\"no_instance_id\\\": true,\\n    \\\"secure_time\\\": true\\n}\", event_log: [EventLog { imr: 0, event_type: 2147483659, digest: \"519245d8b6b54b48b57143e6647d41e5e6f3b4ef10a73... (truncated 6879 bytes)",
    expected_measurements: "ExpectedMeasurements { rtmrs: Measurements { mrtd: [198, 133, 24, 160, 235, 180, 33, 54, 193, 43, 34, 117, 22, 79, 140, 114, 242, 95, 169, 163, 67, 146, 34, 134, 135, 237, 110, 156, 174, 185, 192, 241, 219, 216, 149, 233, 207, 71, 81, 33, 192, 41, 220, 71, 231, 14, 145, 253], rtmr0: [55, 68, 177, 84, 6, 149, 0, 164, 102, 245, 20, 37, 59, 73, 133, 130, 153, 178, 225, 189, 196, 78, 61, 85, 115, 55, 216, 30, 130, 139, 237, 246, 160, 65, 15, 39, 211, 161, 140, 147, 46, 94, 73, 225, 196, 33, 87, 55], rtmr1: [75, 102, 232, 136, 200, 223, 167, 165, 4, 252, 124, 160, 96, 171, 158, 45, 5, 18, 51, 241, 21, 215, 19, 4, 8, 85, 112, 199, 172, 113, 245, 161, 144, 163, 226, 55, 209, 95, 9, 101, 150, 122, 120, 83, 155, 160, 215, 135], rtmr2: [90, 65, 201, 247, 28, 229, 101, 91, 107, 166, 5, 254, 13, 0, 160, 160, 90, 221, 116, 113, 172, 170, 166, 170, 21, 91, 206, 30, 4, 184, 32, 79, 15, 255, 174, 194, 230, 201, 95, 252, 20, 66, 179, 126, 20, 17, 39, 217] }, local_sgx_event_digest: [116, 202, 147, 155, 140, 60, 116, 170, 179, 195, 9, 102, 167, 136, 247, 116, 57, 81, 213, 74, 147, 106, 113, 29, 208, 20, 34, 240, 3, 255, 157, 246, 102, 111, 60, 197, 73, 117, 210, 228, 243, 92, 130, 152, 101, 88, 63, 15], report_data_version: V1 }"
}
```

Comment (DSharifi) [reply attestation/src/attestation.rs:83]: Makes sense. Created a follow up so we can resolve and think more about it:

https://github.com/near/mpc/issues/932

File: test-utils/src/attestation.rs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+pub const TEST_APP_COMPOSE_WITH_SERVICES_STRING: &str =
+    include_str!("../assets/app_compose_with_services.json");
+
+pub const TEST_LAUNCHER_IMAGE_COMPOSE_STRING: &str =
+    include_str!("../assets/launcher_image_compose.yaml");
+
+/// App compose field corresponds to the `DEFAULT_IMAGE_DIGEST` field in
+/// test-utils/assets/launcher_image_compose.yaml
+pub const TEST_MPC_IMAGE_DIGEST_HEX: &str = include_str!("../assets/mpc_image_digest.txt");
+/// sha256sum test-utils/assets/launcher_image_compose.yaml
```
[the remark this answers is not in the document]
Comment (pbeza) [reply test-utils/src/attestation.rs]: Credit to Claude.ai for the massive help with the refactor — including updating the comments! 🤖

**Comment by pbeza** [comment]
Flaky test? 🤔
https://github.com/near/mpc/actions/runs/17159484291/job/48684802487?pr=828

File: test-utils/src/attestation.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (netrome) [reply test-utils/src/attestation.rs]: Good bot :)

**Comment by netrome** [comment]
This PR only adds unit level testing for this behavior right? Because I don't see any pytests modified. I.e. the title is misleading.

I'll update the title. @pbeza let me know if I misunderstood something.

**Comment by pbeza** [comment]
BTW, this PR also resolves https://github.com/near/mpc/issues/846.

**Comment by netrome** [comment]
> > This PR only adds unit level testing for this behavior right? Because I don't see any pytests modified. I.e. the title is misleading.
> > I'll update the title. @pbeza let me know if I misunderstood something.
> 
> @netrome The title was actually correct with "integration test". It's adding integration tests in the `/tests` folder of the contract crate which is defined as as an integration test https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html in Rust. Unit tests are all the tests you usually write in the src file
> 
> If anything i'd argue it's closer to a system test of the contract than a unit test, as we actually build the whole contract and deploy it on a test runtime.
> 
> Edit: To expand on the pytest comment, I'd argue that our pytests are **system test**s, not integration tests. This book from google also goes into some of these topics, but is not Rust specific. https://sre.google/sre-book/testing-reliability/

Nice! You are right. I guess the challenge here is that it's more of a continuous spectrum rather than a discrete distinction. To me system tests is when we test the whole system on a live network, e.g. our devnet setup whereas integration testing is when we're integrating multiple larger components as we do in our pytests. And anything targeting isolated functionality is unit testing, even if it covers the whole contract. But I think it makes sense to call these "integration tests", and perhaps also promoting our pytests to the rank of "system tests".
